storage/src/mmr/hasher.rs

```diff
@@ -1,82 +1,212 @@
 
-impl<'a, H: CHasher> Hasher<'a, H> {
-    pub(crate) fn new(hasher: &'a mut H) -> Self {
+impl<'a, H: CHasher> Standard<'a, H> {
+    /// Creates a new [Standard] hasher.
+    pub fn new(hasher: &'a mut H) -> Self {
         Self { hasher }
     }
 
-    /// Computes the hash for a leaf given its position and the element it represents.
-    pub(crate) fn leaf_hash(&mut self, pos: u64, element: &[u8]) -> H::Digest {
+    pub(crate) fn update_with_pos(&mut self, pos: u64) {
+        self.hasher.update(&pos.to_be_bytes());
+    }
+
+    pub(crate) fn update_with_digest(&mut self, digest: &H::Digest) {
+        self.hasher.update(digest.as_ref());
+    }
+
+    pub(crate) fn update_with_element(&mut self, element: &[u8]) {
+        self.hasher.update(element);
+    }
+
+    pub(crate) fn finalize(&mut self) -> H::Digest {
+        self.hasher.finalize()
+    }
+}
+
+impl<H: CHasher> Hasher<H> for Standard<'_, H> {
+    fn inner(&mut self) -> &mut H {
+        self.hasher
+    }
+
+    async fn leaf_digest(&mut self, pos: u64, element: &[u8]) -> Result<H::Digest, Error> {
         self.update_with_pos(pos);
         self.update_with_element(element);
-        self.finalize_reset()
+        Ok(self.finalize())
     }
 
-    /// Computes the hash for a node given its position and the hashes of its children.
-    pub(crate) fn node_hash(
-        &mut self,
-        pos: u64,
-        left_hash: &H::Digest,
-        right_hash: &H::Digest,
-    ) -> H::Digest {
+    fn node_digest(&mut self, pos: u64, left: &H::Digest, right: &H::Digest) -> H::Digest {
         self.update_with_pos(pos);
-        self.update_with_hash(left_hash);
-        self.update_with_hash(right_hash);
-        self.finalize_reset()
+        self.update_with_digest(left);
+        self.update_with_digest(right);
+        self.finalize()
     }
 
-    /// Computes the root hash for an MMR given its size and an iterator over the hashes of its
-    /// peaks. The iterator should yield the peak hashes in decreasing order of their height.
-    pub(crate) fn root_hash<'b>(
+    fn root_digest<'b>(
         &mut self,
         pos: u64,
-        peak_hashes: impl Iterator<Item = &'b H::Digest>,
+        peak_digests: impl Iterator<Item = &'b H::Digest>,
     ) -> H::Digest {
         self.update_with_pos(pos);
-        for hash in peak_hashes {
-            self.update_with_hash(hash);
+        for digest in peak_digests {
+            self.update_with_digest(digest);
         }
-        self.finalize_reset()
+        self.finalize()
     }
 
-    pub(crate) fn update_with_pos(&mut self, pos: u64) {
-        self.hasher.update(&pos.to_be_bytes());
+    fn digest(&mut self, data: &[u8]) -> H::Digest {
+        self.hasher.update(data);
+        self.finalize()
     }
-    pub(crate) fn update_with_hash(&mut self, hash: &H::Digest) {
-        self.hasher.update(hash.as_ref());
+}
+
+/// Hasher for computing leaf, node and root digests when the tree is being grafted onto another
+/// MMR.
```

Comment:
nit: I think it would be extremely useful to attach any sort of ascii diagram you can think of here.

Reply:
this is a very "novel" idea that takes a few reads to get.

Reply:
Just noticed some of the terminology I defined around this has disappeared, i will re-add it, with a diagram.

Reply:
documentation restored & diagram added.